torch/onnx/verification.py

```diff
@@ -136,136 +73,271 @@ def _run_ort(ort_sess, inputs):
+        # Handle empty kwargs (normally removed by flatten).
+        onnx_inputs = onnx_inputs[:-1]
+    if remained_onnx_input_idx is not None:
+        return [onnx_inputs[i] for i in remained_onnx_input_idx]
+    else:
+        return onnx_inputs
+
+
+def _try_clone_model(model):
+    """Used for preserving original model in case forward mutates model states."""
+    try:
+        return copy.deepcopy(model)
+    except Exception:
+        warnings.warn(
+            "Failed to clone model. Model state might be mutated during verification."
+        )
+        return model
+
+
+def _compare_ort_pytorch_model(
+    model,
+    ort_session,
+    input_args,
+    input_kwargs,
+    additional_test_inputs,
+    remained_onnx_input_idx,
+    flatten,
+    rtol,
+    atol,
+):
+    """Compare outputs from ONNX model runs with outputs from PyTorch model runs.
+
+    ONNXRuntime is used for model execution backend for ONNX model.
+
+    Raises:
+        AssertionError: if outputs from ONNX model and PyTorch model are not
+            equal up to specified precision.
+    """
+
+    def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
+        pt_args, pt_kwargs = _prepare_input_for_pytorch(input_args, input_kwargs)
+        # TODO: remove this and treat mutating model separately. See #77679
+        model_copy = _try_clone_model(model)
+        pt_outs = model_copy(*pt_args, **pt_kwargs)
+
+        ort_inputs = _prepare_input_for_ort(
+            input_args, input_kwargs, remained_onnx_input_idx, flatten
+        )
+        ort_outs = _run_ort(ort_session, ort_inputs)
+
+        _compare_ort_pytorch_outputs(ort_outs, pt_outs, rtol, atol)
+
+    compare_ort_pytorch_model_with_input(input_args, input_kwargs)
+
+    if additional_test_inputs:
+        for test_input_args in additional_test_inputs:
+            compare_ort_pytorch_model_with_input(test_input_args, {})
 
 
 def verify(
     model: Union[torch.nn.Module, torch.jit.ScriptModule],
-    input=None,
-    batch_size=2,
-    rtol=0.001,
-    atol=1e-7,
+    input_args,
+    input_kwargs=None,
     do_constant_folding=True,
     dynamic_axes=None,
-    test_with_inputs=None,
     input_names=None,
     output_names=None,
-    fixed_batch_size=False,
-    dict_check=True,
     training=None,
+    opset_version=None,
+    keep_initializers_as_inputs=True,
+    verbose=False,
+    fixed_batch_size=False,
+    use_external_data=False,
+    additional_test_inputs=None,
     remained_onnx_input_idx=None,
     flatten=True,
-    verbose=False,
     ort_providers=_ORT_PROVIDERS,
-    ort_optim_on=True,
-    use_external_data=False,
-    opset_version: Optional[int] = None,
-    keep_initializers_as_inputs: bool = True,
+    rtol=0.001,
+    atol=1e-7,
+    **kwargs,
 ):
-    """Verify model export to ONNX with ONNXRuntime."""
+    """Verify model export to ONNX with ONNXRuntime.
+
+    Args:
+        model (torch.nn.Module or torch.jit.ScriptModule): See :func:`torch.onnx.export`.
```

Comment:
Can we use type annotations (rather than the docstring) to document the types?
Is this for consumption by sphinx for display on onnx.html? Sphinx supports reading type annotations, but I'm not sure if it's enabled in the PyTorch doc build.

Reply:
Putting the types in the doc string seems unnecessary (and it might get out of sync) if we have type annotations. Any reason to keep both?

Reply:
The docstring looks like style convention. Added type annotations.

Reply:
The convention seems to exist before type annotations are used. I like the idea of keeping only the annotations